[nexus] add `POST /v1/instances/{instance}/force-terminate`

hawkw opened the pull request:
Fixes #4004 

Today, requests to stop a running instance must by necessity involve the
instance's active Propolis (sled agent sends the stop request to
Propolis; the instance is stopped when Propolis says so, at which point
sled agent cleans up the Propolis zone and all related objects). If an
instance's Propolis is not responding, or there is no active Propolis,
there is no obvious way to clean up the instance and recover.

In order to allow resolving otherwise stuck instances, this commit
introduces a new API endpoint, `POST
/v1/instances/{instance}/force-terminate`, which calls directly into the
sled-agent's `instance-ensure-unregistered` API, telling it to rudely
terminate the VMM and destroy the Propolis zone, without waiting for it
to shut down politely.

The one complex-ish bit is the issue I fixed in commit [34c3058] around
what this API does in the case where the sled-agent has forgotten an
instance due to an unexpected restart. The sled-agent
`instance-ensure-unregistered` API is idempotent, which is right and
proper...but it introduces an issue in the case of a forgotten VMM after
a sled-agent restart. The `instance-ensure-unregistered` call will
return `None` because sled-agent doesn't know about that instance, but
if this is the first time we have discovered that sled-agent doesn't
know about the instance, we will need to move it to `Failed`. This is
okay to do, because the VMM generation number guards against the case
where we have raced with _another_ instance-force-terminate call. We
will only move the VMM to `Failed` in that case if no one else has moved
it to `Destroyed` as the result of a successful
`instance-ensure-unregistered` in the interim.

[34c3058]: https://github.com/oxidecomputer/omicron/commit/34c30587f5c145bc4013f97ccaceb34628be5d68

gjcolombo commented on nexus/external-api/src/lib.rs:
```diff
@@ -1166,6 +1166,18 @@ pub trait NexusExternalApi {
         path_params: Path<params::InstancePath>,
     ) -> Result<HttpResponseAccepted<Instance>, HttpError>;
 
+    /// Stop instance, violently.
```
```suggestion
        /// Terminate instance
        ///
        /// Immediately halts a running instance by rudely terminating its
        /// virtual machine process. This immediately moves the instance to the
        /// "stopped" state without transitioning through the "stopping" state.
        /// This operation can be used to recover an instance that is not
        /// responding to requests to stop issued through the instance stop API.
```

The difference between this and regular instance stop is subtle enough that I'm inclined to wordsmith the documentation a bit. The tricky bit is that the two operations look the same to the guest OS, so we can't pitch this as "stop tries to shut down the guest, terminate is a power pull"--it's really more "stop is a power pull, terminate blows away the current running VM state so you can try again from Stopped." Hopefully this will get readers pointed in the right direction.

gjcolombo commented on nexus/src/app/instance.rs:
```diff
@@ -774,6 +774,108 @@ impl super::Nexus {
+            .instance_fetch_with_vmm(opctx, &authz_instance)
+            .await?;
+
+        // Is the instance currently incarnated in a VMM process?
```
It's less critical for getting an instance out of a wedged-in-Starting/wedged-in-Stopping state, but I'd envisioned this also terminating the current migration target, if there is one (so that there are no living VMMs consuming resources on behalf of the instance).

hawkw replied:
Mm, yeah, this is a good point, we should probably fetch the target VMM as well, and unregister it too.

gjcolombo commented on nexus/src/app/instance.rs:
```diff
@@ -774,6 +774,108 @@ impl super::Nexus {
+            // has misplaced this instance. Therefore, we will attempt to mark
+            // the VMM as `Failed` at the generation after which we observed the
+            // VMM. This is safe to do here, because if the instance has been
+            // unregistered due to a race with another
+            // instance-ensure-unregistered request (rather than a sled-agent
```
nit: this is also true if the VMM terminated normally, correct? (I *hope* the same logic holds; if it doesn't I think I've gravely misunderstood something!)

hawkw replied:
Ah, yeah, that's correct.

gjcolombo commented on nexus/src/app/instance.rs:
```diff
@@ -774,6 +774,108 @@ impl super::Nexus {
+        // Ladies and gentlemen, we got him!
+        let propolis_id = PropolisUuid::from_untyped_uuid(vmm.id);
+        let sled_id = SledUuid::from_untyped_uuid(vmm.sled_id);
+        let unregister_result =
+            self.instance_ensure_unregistered(&propolis_id, &sled_id).await;
```
Does the start saga clean up properly if this happens between its `ensure_registered` and `ensure_running` steps? I *think* it works out: `sis_ensure_running` will fail; `sis_ensure_registered_undo` will also fail and try to move the VMM to Failed, but this doesn't block the saga from continuing to unwind; then I think we'll end up in SagaUnwound and end up with a VM that can be started again. Does that sound about right? If so, it might be worthwhile to add a comment mentioning this case.

hawkw replied:
I *think* so, but I would like to figure out whether it's possible to test this...

gjcolombo replied:
Can we do something with error types instead? That is: have `mark_vmm_failed` return its own error enum with "DB query failed" and "update too old" variants, and have `sis_ensure_registered_undo` match on specific error codes from this callee, such that a new failure mode will break the match. WDYT? It'd be nice to have an integration test, too, but I'm similarly having trouble figuring out how to inject a failure into this specific call, since we don't have a CRDB test double that I know of.

hawkw replied:
Hm, actually, the write to CRDB in `mark_vmm_failed` is done by a call to `vmm_update_runtime`, which returns `Ok(false)` if the VMM exists but wasn't updated (e.g. if the generation has advanced). So, we don't even hit the error path (which gets ignored anyway) in `mark_vmm_failed` in that case...

hawkw replied:
Which, now that I look at it, indicates a tiny possible inefficiency in `mark_vmm_failed` --- currently, we try to run an update saga even if the VMM did not move to failed (because it was already failed/destroyed), which means we will probably start spurious update sagas there. I'll clean that up.

hawkw replied:
I think I'm probably fine with having `mark_vmm_failed` stick with the convention of returning `Ok(false)` if the generation has changed, instead of an error variant, since it seems like we generally follow that convention for similar code. On one hand, I do have a sort of ideological preference for an `Ok(_)` to _always_ mean "yes, we actually wrote the desired update to CRDB", but on the other hand, a lot of saga undo actions probably just `?` these functions, and would probably prefer to get an `Ok` in any case that doesn't mean they should unwind. I dunno.

hawkw replied:
`mark_vmm_failed` actually has the type signature of `async fn(...) -> Result<(), Error>` but there are actually no conditions in which it will ever currently return an error. That's cool.

I might just change the type signature to not return `Result`

hawkw replied:
Upon further perusal of `sis_ensure_registered_undo`, I noticed that it was actually handling sled-agent unregister errors in a pretty outdated way: it was treating `vmm_gone` errors as a probable failure, and just ignoring any other errors communicating with the sled-agent, as described in this comment:

https://github.com/oxidecomputer/omicron/blob/0640bb277df110f3e740464bf4bacdf2bd24c897/nexus/src/app/sagas/instance_start.rs#L673-L691

This predates the changes to `vmm_gone`/`instance_unhealthy` semantics from RFD 486/#6455, and isn't quite correct. I've changed it to make the "other error from sled agent" be the case that fails the saga, although we should probably retry communication errors eventually. 

The changes to `mark_vmm_failed` and `sis_ensure_registered_undo` are in a923f2a6a058fc1b42da33c8d1f52da6592b0bd1. Arguably, it's not *really* in scope for this PR at that point; I'd be happy to cherry-pick it out to a separate branch if you think that's worth doing?

hawkw commented:
@gjcolombo I've changed this so that we now also tear down migration target VMMs (see d6a2bd47d3051830093b3f5e114068a4267209dc). 

In the process, I also changed the code for terminating a VMM to also synchronously attempt to run an update saga to completion (e841bd0af3b369bc1d316375f0f0a28848b44289). This came up because I noticed that, when force-terminating a `Migrating` instance, the returned instance will still appear to be `Migrating` even though both VMMs have been destroyed, because it still has a migration ID set until an update saga removes it. We had decided to do that to stop migrating VMMs from briefly appearing to be `Stopped` when a migration completes and an active VMM is destroyed, but the instance hasn't yet been updated to reflect that it is now running on the target. I think this is still correct behavior, but it seemed kinda sad to "force terminate" an instance and see that the force-terminate succeeded but it's still `Migrating`, so I made the force-terminate method run the whole saga.

gjcolombo commented on nexus/src/app/instance.rs:
```diff
@@ -818,35 +869,70 @@ impl super::Nexus {
-                    "asked to force terminate an instance that was already \
-                     unregistered";
+                    "asked to force terminate an instance's {vmm_role} VMM ;
+                     thatwas already unregistered";
```
```suggestion
                     that was already unregistered";
```

gjcolombo commented on nexus/src/app/instance.rs:
```diff
@@ -818,35 +869,70 @@ impl super::Nexus {
+                // unset, and it seems a bit sad to return a "Migrating"
+                // instance to a caller who tries to force-kill it.
+                //
+                // TODO(eliza): in the case where we are terminating both an
+                // active VMM and migration target, we will unregister them in
+                // sequence, running separate update sagas for both VMMs being
+                // terminated. It would be a bit more efficient to terminate
+                // both VMMs, update CRDB, and then run a *single* update saga
+                // to process both VMMs being destroyed. However, this requires
+                // a bit of annoying refactoring to existing APIs, and I'm not
+                // sure if improving the `instance-force-terminate` endpoint's
+                // latency is particularly important...
```
I could be missing something, but it seems like we could at least return the generated `saga` from this function and then have the caller actually *execute* only one of the prepared sagas. (It'd be even better to call `update_saga_needed` just once, after applying both VMM changes, but I agree that that would be a bigger refactoring.)

hawkw replied:
Done in https://github.com/oxidecomputer/omicron/pull/6795/commits/d69a1ceb3d9bd342a547747b4b5cfb3c52d6f2e8 --- can't believe I didn't think of that.

iximeow commented on nexus/tests/integration_tests/instances.rs:
```diff
@@ -6028,6 +6027,222 @@ async fn test_instance_v2p_mappings(cptestctx: &ControlPlaneTestContext) {
+    );
+    assert_eq!(instance.runtime.run_state, InstanceState::Stopped);
+
+    // Okay, what if the instance is already gone?
+    let instance = dbg!(
+        instance_post(&client, &already_gone_name, InstanceOp::ForceTerminate)
+            .await
+    );
+    // This time, the instance will go to `Failed` rather than `Stopped` since
+    // sled-agent is no longer aware of it.
+    assert_eq!(instance.runtime.run_state, InstanceState::Failed);
```
fwiw this initially seemed concerning ("if you force-stop an instance in the right state, now you have a failed instance?"), but i think i've figured out the preconditions by reading more carefully: this is specifically if you force-stop an instance where Nexus believes a sled-agent is responsible for the instance, but that sled-agent doesn't know it exists, and that _already_ is a sign that something has gone wrong. so it's an instance that probably should be failed anyway, we just hadn't figured that out yet?

if it helps on wording, i think "already gone" is what primed me to think this might be applicable in some kind of race around stopping.

that said, i _think_ there's a race where if  you force-stop an instance alongside a call to `instance_stop` you can end up with an instance in Failed where auto-restart would just turn it back on again? specifically if `instance_stop` has reached out to sled-agent, sled-agent responds that it has stopped the instance, but we haven't actually recorded that state update yet, then `force_stop` could show up and see Nexus thinks that sled-agent has the VMM while that sled-agent thinks it's gone.

i haven't looked at the `instance_stop` code much yet so maybe i'm misunderstanding and there's not actually something there..? but if i'm following these bits right, it seems kind of unfortunate that trying too enthusiastically to stop an instance could end up making it auto-restart.

hawkw replied:
Multiple racing stop attempts shouldn't transition an instance to failed, since the query in `mark_vmm_failed` that transitions the VMM to failed is conditional on the sequence number of the VMM record remaining the same as when the VMM was read. If the VMM was destroyed between when we read the record initially and when we asked the sled-agent to destroy it, the sequence number will have advanced, and we won't move it to failed. It only moves to failed if it was already nonexistent.

I see your point about auto-restarts, though: perhaps we should just always move a force-stopped instance to `Stopped` when encountering a sled-agent error that would otherwise move it to `Failed`, since you're trying to stop it anyway, which presumably means you have other plans for the instance? I dunno. @gjcolombo, what do you think?

hawkw replied:
It occurs to me that we should probably also bail out early if we see the VMM record is already `Destroyed`, though, rather than asking the sled-agent to destroy it again...

iximeow replied:
ok! i see your point about the non-raciness now - if there's still a Propolis ID for an instance, the `force-terminate` call might cause Nexus to get to `instance_ensure_unregistered` and in fact ask `sled-agent` to ensure the Propolis has been terminated, but in that case `sled-agent`'s `ensure_unregistered` would return `VmmUnregisterResponse { updated_runtime: None }` rather than an error that the Propolis that should be unregistered is already gone. to the point of sled-agent docs, `ensure_unregistered` is dempotent.

so, among other things, new `Err(InstanceStateChangeError::SledAgent(e)) if e.vmm_gone()` arm in `instance.rs` _shouldn't_ ever be reached, and _that_ case of marking an instance failed doesn't matter so much.

fully convinced it's not racy now, thanks for bearing with me.

iximeow commented on nexus/external-api/src/lib.rs:
```diff
@@ -1166,6 +1166,24 @@ pub trait NexusExternalApi {
+    ///
+    /// Immediately halts a running instance by rudely terminating its
+    /// virtual machine process. This immediately moves the instance to the
+    /// "stopped" state without transitioning through the "stopping" state.
+    /// This operation can be used to recover an instance that is not
+    /// responding to requests to stop issued through the instance stop API.
```
putting my mildly-uninformed-user hat on, is there something important i could be missing out on by not transitioning through "stopping"? resources that could be leaked (seems unlikely), or internal instance state that could get weird (seems more likely)? from what's here it doesn't seem unreasonable for a user to always `/force-terminate` on the assumption that it's more like yanking power, and i dunno how much anyone would be disturbed by that. i recognize this is also kind of the ambiguity @gjcolombo was trying to address, sorry :)

putting my Oxide engineer hat on, it feels like any reason to use `/force-terminate` is a result of a user trying to unwedge themselves from an Oxide bug. so maybe that's the kind of warding this documentation deserves? though i'm still not sure how load-bearing `stopping` is..

hawkw replied:
The instance will still transition through the `Stopping` state if you are querying it in the meantime, which will be visible in e.g. the console. It's just that the `force-terminate` API endpoint does not _return_ until the instance has advanced to `Stopped`.

hawkw commented:
Turning this into a draft, as we've recently come to the conclusion that we might be better off not having this kind of API at all: https://github.com/oxidecomputer/omicron/issues/4004#issuecomment-2403347700

hawkw commented:
This is obsoleted by PR #7548, which gives `sled-agent` a way to violently terminate stuck Propolis zones without requiring a separate API endpoint. Merging that PR will fix #4004 and close this PR.
